cli_tools/common/distro/distro.go

```diff
@@ -0,0 +1,187 @@
+func FromLibguestfs(distro string, major string, minor string) (r Release, e error) {
+	distro = strings.ToLower(distro)
+	if distro == windows {
+		// We don't currently need Windows for this parsing, and punting
+		// since it's not immediately clear how to represent their
+		// version strings, as the user-facing value (Windows 2008r2)
+		// does not match what's used internally by our tools (NT 5.2)
+		// https://en.wikipedia.org/wiki/List_of_Microsoft_Windows_versions
+		return r, errors.New("Windows not yet implemented")
+	}
+	if distro == "" || major == "" {
+		return nil, errors.New("distro and major are required")
+	}
+	return release{
+		distro: distro,
+		major:  major,
+		minor:  minor,
+	}, nil
+}
+
+// ParseGcloudOsParam parses the value associated with the `--os` parameter
+// from the `gcloud compute image` tools.
+//
+// https://cloud.google.com/sdk/gcloud/reference/compute/images/import#--os
+func ParseGcloudOsParam(osFlagValue string) (r Release, e error) {
+	os := strings.ToLower(osFlagValue)
+	if strings.HasSuffix(os, "-byol") {
+		os = strings.TrimSuffix(os, "-byol")
+	}
+	if strings.HasPrefix(os, windows) {
+		// We don't currently need Windows for this parsing, and punting
+		// since it's not immediately clear how to represent their
+		// version strings, as the user-facing value (Windows 2008r2)
+		// does not match what's used internally by our tools (NT 5.2)
+		// https://en.wikipedia.org/wiki/List_of_Microsoft_Windows_versions
+		return r, errors.New("Windows not yet implemented")
+	}
+	parts := strings.Split(os, "-")
+	distro := parts[0]
+
+	// 1. Format is `(debian|centos|rhel|opensuse)-major`.
+	//   Canonical example: debian-8
+	if distro == debian ||
+		distro == centos ||
+		distro == rhel ||
+		distro == opensuse {
+		if len(parts) != 2 {
+			return r, fmt.Errorf("expected pattern of `distro-version`. Actual: `%s`", os)
+		}
+
+		return release{
+			distro: distro,
+			major:  parts[1],
+		}, nil
+	}
+
+	// 2. Format is `sles(-variant)?-major`.
+	//   Canonical examples: sles-sap-12, sles-12
+	if distro == sles {
+		switch len(parts) {
+		case 2:
+			return release{
+				distro: distro,
+				major:  parts[1],
+			}, nil
+		case 3:
+			return release{
+				distro:  distro,
+				variant: parts[1],
+				major:   parts[2],
+			}, nil
+		default:
+			return r, fmt.Errorf("unrecognized SLES identifier: `%s`", os)
+		}
+	}
+
+	// 3. Format is `ubuntu-(major)(minor)`.
+	//   Canonical example: ubuntu-1804
+	if distro == ubuntu {
+		if len(parts) != 2 {
+			return r, fmt.Errorf("expected pattern of `distro-version`. Actual: `%s`", os)
+		}
+
+		version := parts[1]
+		// We support importing LTS versions, which end in 04.
+		if len(version) != 4 {
+			return r, fmt.Errorf("expected version with length four. Actual: `%s`", version)
+		}
+
+		return release{
+			distro: distro,
+			major:  version[:2],
+			minor:  version[2:],
+		}, nil
+	}
+
+	return r, fmt.Errorf("unrecognized identifier: `%s`", os)
```

Comment:
Is there any chance to use this error? It seems that the caller swallowed it.

Reply:
In the current usage, you're right, it gets swallowed. That happens since we've already validated the `--os` argument parameter further up in the method.